File: Module1 - Zachary Abbe -CS499/Project1-ZacharyAbbe-CS320/Appointment.java
/***************
 * Name: Zachary Abbe
 * Date: 04/12/2025
 * Desc: This is the Appointment Class that houses all the variables for the object
 **************/
import java.util.Date;

public class Appointment {

    private String uniqId;
    private Date date;
    private String description;


     // validates the uniqueID

    private final boolean validateID(String uniqueID) {
        if(uniqueID == null || uniqueID.length() > 10) {
            return false;
        }
        return true;
    }


     // validates the Name

    private final boolean validateDate(Date _date) {

        if(_date == null || _date.before(new Date())) {
            return false;
        }
        return true;
    }

     //validates the description

    private final boolean validateDescription(String description) {
        if(description == null || description.length() > 50 || description.equals("")) {
            return false;
        }
        return true;
    }

    // constructor for the object
    public Appointment(String uniqueID,
                       Date date,
                       String description) {

        if(!this.validateID(uniqueID)) {
            throw new IllegalArgumentException("Invalid ID");
        }

        if(!this.validateDate(date)) {
            throw new IllegalArgumentException("Invalid date");
        }

        if(!this.validateDescription(description)) {
            throw new IllegalArgumentException("Invalid description");
        }


        setID(uniqueID);
        setDate(date);
        setDescription(description);

    }
    //getters and setters
    public String getUniqueID() {
        return uniqId;
    }


    private void setID(String uniqueID){
        this.uniqId = uniqueID;
    }


    public Date getDate() {
        return date;
    }

    public void setDate(Date date) {
        if(!this.validateDate(date)) {
            throw new IllegalArgumentException("Invalid date");
        }
        this.date = date;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        if(!this.validateDescription(description)) {
            throw new IllegalArgumentException("Invalid description");
        }
        this.description = description;
    }
}

File: CS499Project1and2-ZacharyAbbe/Appointment.cpp
/******************************************************************************
*Name: Zachary Abbe
*Date: 07/16/2025
*Desc: This is the Appointment.cpp that will be called by main to perform our functions
******************************************************************************/

#include "Appointment.h"
#include <stdexcept>

using namespace std;

const int IDlength = 10;
const int descLength = 50;

// this uses the Appointment class function validateID and provides the code that is ran when this is called
bool Appointment::validateID(const string& uniqueID) 
{
	return !uniqueID.empty() && uniqueID.length() <= IDlength;
}

bool Appointment::validateDate(const chrono::system_clock::time_point& _date) 
{
	return _date > chrono::system_clock::now();
}

bool Appointment::validateDescription(const string& description) 
{
	return !description.empty() && description.length() <= descLength;
}
Appointment::Appointment(const string& uniqueID, const chrono::system_clock::time_point& date, const string& description)
{
	if (!validateID(uniqueID))
	{
		throw invalid_argument("Invalid ID");
	}

	if (!validateDate(date))
	{
		throw invalid_argument("Invalid Date");
	}

	if (!validateDescription(description))
	{
		throw invalid_argument("Invalid Description");
	}

	this->uniqId = uniqueID;
	this->date = date;
	this->description = description;
 }

//getters for the object
string Appointment::getUniqueID() const 
{
	return uniqId;
}

chrono::system_clock::time_point Appointment::getDate() const 
{
	return date;
}

string Appointment::getDescription() const 
{
	return description;
}

//setters for the object

void Appointment::setDate(const chrono::system_clock::time_point& date) 
{
	if (!validateDate(date))
	{
		throw invalid_argument("Invalid Date");
	}
	this->date = date;
}

void Appointment::setUniqueID(const string& uniqueID) 
{
	if (!validateID(uniqueID))
	{
		throw invalid_argument("Invalid ID");
	}
	this->uniqId = uniqueID;
}
void Appointment::setDescription(const string& description)
{
	if (!validateDescription(description))
	{
		throw invalid_argument("Invalid Description");
	}
	this->description = description;
}